### ISSUE TITLE
Run lint on save not working

### ISSUE THREAD

**Issue by skhaz** [zed-industries/zed#9279]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

I have Black installed in my Python virtual environment, not globally, and I would like Zed to use the Black from the virtual environment without having to force me to install it globally.

The configuration below does not work; it only works if I run it through the terminal.

How can I solve this, or is it a bug?

```json
{
  "features": {
    "copilot": false
  },
  "theme": "One Dark",
  "language_overrides": {
    "Python": {
      "format_on_save": {
        "external": {
          "command": "black",
          "arguments": ["-"]
        }
      }
    }
  }
}
```

### Environment

Zed: v0.125.4 (Zed)
OS: macOS 14.3.1
Memory: 16 GiB
Architecture: aarch64

**Comment by Moshyfawn** [comment]
You may be interested in both #6083 and this Zed Python [setup guide](https://github.com/zed-industries/zed/blob/main/docs/src/languages/python.md).

**Comment by skhaz** [comment]
I have read that issue @Moshyfawn, and I have this file on my project. It does not work either.

**Comment by sakisv** [comment]
Same situation with ruff, I've installed it in a venv and used this configuration:

```
  "language_overrides": {
    "Python": {
      "format_on_save": {
        "external": {
          "command": "ruff",
          "arguments": [
            "check",
            "--exit-zero",
            "--fix",
            "--stdin-filename",
            "{buffer_path}",
            "-"
          ]
        }
      }
    }
  ```

which fails with this error in the logs:

```
2024-05-10T00:52:10+03:00 [ERROR] crates/editor/src/editor.rs:8462: failed to format via external command "ruff"

Caused by:
    No such file or directory (os error 2)
```

**Comment by skhaz** [comment]
I am planning to use ruff soon, this should be fixed.

**Comment by sakisv** [comment]
Actually I just noticed that you could kinda get around this limitation for now by using project-specific settings[^1]. 

Create a `.zed/settings.json` at the root of your project and then you can provide the path to the command in your venv, like so:

```
{
  "language_overrides": {
    "Python": {
      "format_on_save": {
        "external": {
          "command": "venv/bin/ruff",
          "arguments": [
            "check",
            "--exit-zero",
            "--fix",
            "--stdin-filename",
            "{buffer_path}",
            "-"
          ]
        }
      }
    }
  }
```


[^1]: https://zed.dev/docs/configuring-zed#folder-specific-settings

**Comment by benbenbang** [comment]
This should be under `language` instead of `language_overrides`, as it will be deprecated if I understand correctly.

**Comment by unickname3** [comment]
Yeah, thats working

```
 "languages": {
    "Python": {
      "formatter": {
        "external": {
          "command": "ruff",
          "arguments": ["format", "-"]
        }
      },
      "format_on_save": "on"
    }
  }
```

**Comment by skhaz** [comment]
It will use Ruff from the current venv?

**Comment by skhaz** [comment]
Another thing, those configs should be per project, not globally. I work on projects that use black & isort and other it just use ruff.

**Comment by benbenbang** [comment]
> It will use Ruff from the current venv?

@skhaz, if the version doesn't matter to you, i.e, one ruff binary rules all
you can use `pipx`, for example, and use it everywhere

**Comment by skhaz** [comment]
Python's ecosystem is SUPER fragmented. I do not want to use pipx for example. I use uv + venv

**Comment by benbenbang** [comment]
that case, just set your `ruff` path in the project `.zed/settings.json`

see the [docs](https://zed.dev/docs/configuring-zed)

**Comment by skhaz** [comment]
It cannot be hardcoded, Zed or the Python plugin should detected the virtual environment and add it to the path.

I have around 20 projects, each one has its own black or ruff inside the venv and I cannot use the same for all.

**Comment by benbenbang** [comment]
Well, sublime text and vscode are used the same way.

Personally, I wrote a small cli to detect the nearest `pyproject.toml` or `.venv` and automatically generate the settings.

**Comment by skhaz** [comment]
Are you sure? On VSCode it picks the black/ruff from the venv. I never intalled it globally.

**Comment by benbenbang** [comment]
I meant per project, not globally. In VSCode, you can select `venv`, so you don't need to do anything else, that's for sure.

**Comment by skhaz** [comment]
Yes, that was my question, whether Zed has the same feature as VSCode.

I am aware that Zed is a new project with a much smaller team compared to Microsoft's team for VSCode.
